packages/vm/src/context.rs

```diff
@@ -359,43 +360,44 @@ fn create_unmanaged_context_data<S: Storage, Q: Querier>() -> *mut c_void {
         storage: None,
         querier: None,
         #[cfg(feature = "iterator")]
-        iter: HashMap::new(),
+        iterators: HashMap::new(),
     };
-    let state = Box::new(data);
-    Box::into_raw(state) as *mut c_void
+    let heap_data = Box::new(data); // move from stack to heap
+    Box::into_raw(heap_data) as *mut c_void // give up ownership
 }
 
 fn destroy_unmanaged_context_data<S: Storage, Q: Querier>(ptr: *mut c_void) {
     if !ptr.is_null() {
-        let mut dead = unsafe { get_data::<S, Q>(ptr) };
-        // ensure the iterator (if any) is dropped before the storage
-        free_iterator(&mut dead);
+        // obtain ownership and drop instance of ContextData when box gets out of scope
+        let mut dying = unsafe { Box::from_raw(ptr as *mut ContextData<S, Q>) };
+        // Ensure all iterators are dropped before the storage
+        destroy_iterators(&mut dying);
     }
 }
 
-unsafe fn get_data<S: Storage, Q: Querier>(ptr: *mut c_void) -> Box<ContextData<S, Q>> {
-    Box::from_raw(ptr as *mut ContextData<S, Q>)
+/// Get a mutable reference to the context's data. Ownership remains in the Context.
+fn get_context_data<S: Storage, Q: Querier>(ctx: &mut Ctx) -> &mut ContextData<S, Q> {
+    let owned = unsafe {
+        Box::from_raw(ctx.data as *mut ContextData<S, Q>) // obtain ownership
+    };
+    Box::leak(owned) // give up ownership
```

Comment:
Interesting one.
https://doc.rust-lang.org/std/boxed/struct.Box.html#method.leak

Should we annotate the lifetime on the return value? Like `&'static mut`?

Reply:
If I understand https://doc.rust-lang.org/nomicon/lifetime-elision.html correctly, then the above signature is a shorthand for

```rust
fn get_context_data<'a, S: Storage, Q: Querier>(ctx: &'a mut Ctx) -> &'a mut ContextData<S, Q> {
```

In any case, the reference to `ContextData` must not live longer than the reference to `Ctx`. So static would be wrong, right?.